### PR TITLE
Add "time-min-milliseconds" rule

### PR DESCRIPTION
**Pull request by hudochenkov**
> Which issue, if any, is this issue related to?

#2249

> Is there anything in the PR that needs further explanation?

No, it's self explanatory.

### CODE REVIEW

File: lib/rules/time-min-milliseconds/__tests__/index.js [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+const rule = rules[ruleName]
+
+const MIN_VALUE = 100
+
+testRule(rule, {
+  ruleName,
+  config: [MIN_VALUE],
+
+  accept: [ {
+    code: "a { transition-delay: 0.2s; }",
```
Comment (jeddy3) [inline lib/rules/time-min-milliseconds/__tests__/index.js]: Can we get a test for `0.1s` please?

File: lib/rules/time-min-milliseconds/__tests__/index.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+  }, {
+    code: "a { -webkit-transition-delay: 0.2s; }",
+  }, {
+    code: "a { tRaNsItIoN-dElAy: 0.2s; }",
+  }, {
+    code: "a { TRANSITION-DELAY: 0.2s; }",
+  }, {
+    code: "a { transition-delay: 3s; }",
+  }, {
+    code: "a { transition-delay: 200ms; }",
```
Comment (jeddy3) [inline lib/rules/time-min-milliseconds/__tests__/index.js]: Can we get a test for `100ms` please?

File: lib/rules/time-min-milliseconds/__tests__/index.js (hunk 1 above)
Comment (hudochenkov) [inline lib/rules/time-min-milliseconds/__tests__/index.js]: Great catch! It was failing for `0.1s` and `100ms` :)

File: docs/user-guide/rules.md [hunk 3]
```diff
@@ -62,6 +62,7 @@ Here are all the rules within stylelint, grouped by the [*thing*](http://apps.wo
 
 ### Time
 
+-   [`time-min-milliseconds`](../../lib/rules/time-min-milliseconds/README.md): Limit the minimum number of milliseconds for time values.
```
Comment (davidtheclark) [inline docs/user-guide/rules.md]: I think "Specify" is a better word here than "Limit": "Specify the minimum number ..."

File: lib/rules/time-min-milliseconds/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+a { animation: slip-n-slide 150ms linear; }
+/**                         ↑
+ *                  This time */
+```
+
+This rule checks `transition-duration`, `transition-delay`, `animation-duration`, `animation-delay`, and those times as they manifest in the `transition` and `animation` shorthands.
+
+## Options
+
+`int`: Minimum number milliseconds for time values.
```
Comment (davidtheclark) [inline lib/rules/time-min-milliseconds/README.md]: number *of* milliseconds

File: lib/rules/time-min-milliseconds/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+```css
+a { transition-duration: 0.08s; }
+```
+
+```css
+a { transition: background-color 6ms linear; }
+```
+
+```css
+a { animation: horse-dance 1s linear 0.01s; }
```
Comment (davidtheclark) [inline lib/rules/time-min-milliseconds/README.md]: I find this to be an ambiguous example. A reader might reasonably come away thinking both `1s` and `0.01s` are not ok. Let's just have one time value in each example, since no text specifies which particular value is the violator.

Same applies to the example below.

File: lib/rules/time-min-milliseconds/__tests__/index.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+  }, {
+    code: "a { animation-delay: 3s; }",
+  }, {
+    code: "a { animation-delay: 200ms; }",
+  }, {
+    code: "a { animation-delay: 5000ms; }",
+  }, {
+    code: "a { animation-duration: 0.15s; }",
+  }, {
+    code: "a { -webkitanimation-duration: 0.15s; }",
```
Comment (davidtheclark) [inline lib/rules/time-min-milliseconds/__tests__/index.js]: Missing hyphen, I think.

File: lib/rules/time-min-milliseconds/__tests__/index.js [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+  }, {
+    code: "a { animation-duration: 4700ms; }",
+  }, {
+    code: "a { animation: foo 0.8s linear; }",
+  }, {
+    code: "a { -webkit-animation: foo 0.8s linear; }",
+  }, {
+    code: "a { animation: foo 0.8s 200ms ease-in-out; }",
+  }, {
+    code: "a { animation-delay: -2.5s; }",
```
Comment (davidtheclark) [inline lib/rules/time-min-milliseconds/__tests__/index.js]: This is a little confusing and needs to be clarified in the docs. Do *any* negative values work? Or is the "minimum" actually an *absolute* value applying on both sides of zero?

File: lib/rules/time-min-milliseconds/index.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+const ruleMessages = require("../../utils/ruleMessages")
+const validateOptions = require("../../utils/validateOptions")
+const keywordSets = require("../../reference/keywordSets")
+const postcss = require("postcss")
+const valueParser = require("postcss-value-parser")
+
+const ruleName = "time-min-milliseconds"
+
+const messages = ruleMessages(ruleName, {
+  expected: time => `Expected no less than ${time} milliseconds`,
```
Comment (davidtheclark) [inline lib/rules/time-min-milliseconds/index.js]: Let's say `Expected of a minimum of ${time} milliseconds` to avoid quarrels about "less" vs "fewer".

File: lib/rules/time-min-milliseconds/index.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+      actual: minimum,
+      possible: _.isNumber,
+    })
+    if (!validOptions) {
+      return
+    }
+
+    root.walkDecls(decl => {
+      if (keywordSets.longhandTimeProperties.has(postcss.vendor.unprefixed(decl.prop.toLowerCase()))) {
+        if (isNotAcceptableTime(decl.value)) {
```
Comment (davidtheclark) [inline lib/rules/time-min-milliseconds/index.js]: These two conditions could be combined into one if statement.

File: lib/rules/time-min-milliseconds/index.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+      if (parsedTime.unit.toLowerCase() === "s" && absoluteTime * 1000 < minimum) {
+        return true
+      }
+
+      return false
+    }
+
+    function complain(decl) {
+      const offset = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0
```
Comment (davidtheclark) [inline lib/rules/time-min-milliseconds/index.js]: Ugh this is awkward. Can you use an explicit argument instead?

**Comment by hudochenkov** [comment]
@davidtheclark Thank you for an excellent review. To be honest, I just took an old rule and changed few things to be a new rule ¯﻿\﻿_﻿(﻿ツ﻿)﻿_﻿/﻿¯ Will cleanup things.

File: lib/rules/time-min-milliseconds/__tests__/index.js (hunk 7 above)
Comment (hudochenkov) [inline lib/rules/time-min-milliseconds/__tests__/index.js]: Good question! I think we should check only positive values.

According to specs `transition-duration` and `animation-duration` can't be negative. Negative values can be in shorthand properties `transition` and `animation`, but it's hard to parse what time is which in shorthands.

Also, negative delay is a rare case. Developer could set negative value in two cases: intentionally for some reason, or by mistake.

What do you think?

File: lib/rules/time-min-milliseconds/index.js [hunk 11]
```diff
@@ -12,7 +12,7 @@ const valueParser = require("postcss-value-parser")
 const ruleName = "time-min-milliseconds"
 
 const messages = ruleMessages(ruleName, {
-  expected: time => `Expected no less than ${time} milliseconds`,
+  expected: time => `Expected of a minimum of ${time} milliseconds`,
```
Comment (davidtheclark) [inline lib/rules/time-min-milliseconds/index.js]: Cut the first "of" here: `Expected a minimum of ${time} ...`

Comment (hudochenkov) [inline lib/rules/time-min-milliseconds/index.js]: Done.

**Comment by hudochenkov** [comment]
There is an important question lost in outdated diff comments. What numbers should this rule check and [my opinion about it](https://github.com/stylelint/stylelint/pull/2289#discussion_r98077424).

File: lib/rules/time-min-milliseconds/__tests__/index.js (hunk 7 above)
Comment (davidtheclark) [inline lib/rules/time-min-milliseconds/__tests__/index.js]: Yes, I agree: if `transition-duration` and `animation-duration` cannot be negative, then let's not use the absolute value here, and only check positive values.

**Comment by davidtheclark** [comment]
@hudochenkov I responded above, then realized it might be lost in the diff, too:

> Yes, I agree: if transition-duration and animation-duration cannot be negative, then let's not use the absolute value here, and only check positive values.

**Comment by davidtheclark** [comment]
Added to Changelog:

> -   Added: `time-min-milliseconds` rule, to replace `time-no-imperceptible` ([#2289](https://github.com/stylelint/stylelint/pull/2289)).
